packages/front-end/components/MetricAnalysis/HistogramGraph.tsx

```diff
@@ -0,0 +1,260 @@
+}
+
+function getTooltipDataFromDatapoint(
+  datapoint: Datapoint,
+  data: Datapoint[],
+  innerWidth: number,
+  yScale: ScaleLinear<unknown, unknown, never>
+) {
+  const index = data.indexOf(datapoint);
+  if (index === -1) {
+    return null;
+  }
+  const x =
+    (data.length > 0 ? (index + 0.5) / (data.length + 1) : 0) * innerWidth;
+  const y = (yScale(datapoint.units ?? 0) ?? 0) as number;
+  return { x, y, d: datapoint };
+}
+
+function getTooltipContents(
+  d: TooltipData,
+  formatter: (value: number, options?: Intl.NumberFormatOptions) => string
+) {
+  return (
+    <>
+      <div className={`mb-2 ${styles.val}`}>n: {d.d.units}</div>
+      <div className="small">
+        <span className="d-inline-block" style={{ width: 40 }}>
+          start:
+        </span>{" "}
+        {formatter(d.d.start)}
+      </div>
+      <div className="small">
+        <span className="d-inline-block" style={{ width: 40 }}>
+          end:
+        </span>{" "}
+        {formatter(d.d.end)}
+      </div>
+    </>
+  );
+}
+
+const HistogramGraph: FC<HistogramGraphProps> = ({
+  data,
+  formatter,
+  height = 220,
+  margin = [15, 15, 30, 80],
+}: HistogramGraphProps) => {
+  const { containerRef, containerBounds } = useTooltipInPortal({
+    scroll: true,
+    detectBounds: true,
+  });
+
+  const [marginTop, marginRight, marginBottom, marginLeft] = margin;
+
+  const width = (containerBounds?.width || 0) + marginRight + marginLeft;
+  const yMax = height - marginTop - marginBottom;
+  const xMax = containerBounds?.width || 0;
+  const graphHeight = yMax;
+  const binWidth = xMax / data.length;
+
+  const xScale = scaleLinear({
+    domain: [0, data.length],
+    range: [0, xMax],
+  });
+
+  const yScale = scaleLinear({
+    domain: [0, Math.max(...data.map((d) => d.units))],
+    range: [yMax, 0],
+  });
+
+  const {
+    showTooltip,
+    hideTooltip,
+    tooltipOpen,
+    tooltipData,
+    tooltipLeft = 0,
+    tooltipTop = 0,
+  } = useTooltip<TooltipData>();
+
+  const [hoverBin, setHoverBin] = useState<number | null>(null);
+  const onHover = (bin: number | null) => {
+    setHoverBin(bin);
+  };
+
+  useEffect(
+    () => {
+      if (hoverBin === null) {
+        hideTooltip();
+        return;
+      }
+      const datapoint = data[hoverBin];
+      if (!datapoint) {
+        hideTooltip();
+        return;
+      }
+      const innerWidth =
+        width - marginLeft - marginRight + width / data.length - 1;
+      const tooltipData = getTooltipDataFromDatapoint(
+        datapoint,
+        data,
+        innerWidth,
+        yScale
+      );
+      if (!tooltipData) {
+        hideTooltip();
+        return;
+      }
+
+      showTooltip({
+        tooltipLeft: tooltipData.x,
+        tooltipTop: tooltipData.y,
+        tooltipData: tooltipData,
+      });
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [hoverBin, data, marginLeft, marginRight, width, showTooltip, hideTooltip]
```

Comment:
i think that this can be pulled out of a use effect into a function that runs when hoverBin changes.

Reply:
i think we should keep the deps. the tooltip (position, content, hovered bin) will need to recalc if any of these things change. Data, size, etc.